### PR TITLE
LUN-201: Refactor fetching logic and AddCustomDisciplineScreen

### PR DESCRIPTION
**Pull request by klinzo**
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

### CODE REVIEW

**Comment by klinzo** [comment]
Also fixes short showing of error when adding a custom discipline.

**Comment by klinzo** [comment]
@SteffiStuffel it would now be really easy to replace the hoc `withCustomDisciplines` with a equivalent hook or function :)

File: src/hooks/useLoadDocuments.ts [hunk 1]
```diff
@@ -17,23 +17,24 @@ export interface DocumentTypeFromServer {
...
   const url = ENDPOINTS.documents.replace(':id', `${discipline.id}`)
-  const documents: ReturnType<DocumentTypeFromServer[]> = useLoadFromEndpoint(url, discipline.apiKey)
-  return transformArticle(documents)
+
+  const response = await getFromEndpoint<DocumentTypeFromServer[]>(url, discipline.apiKey)
+  return formatServerResponse(response)
 }
 
+const useLoadDocuments = (discipline: Discipline): ReturnType<DocumentType[]> =>
+  useLoadFromEndpoint(loadDocuments, discipline)
```
Comment (SteffiStuffel) [inline src/hooks/useLoadDocuments.ts:38]: Is this "PingPong" really needed?
-> In useLoadDocuments call method from useLoadFromEndpoint
-> in useLoadFromEndpoint call method (loadDocuments) from useLoadDocuments
-> In useLoadDocuments (loadDocuments-method) call method (getFromEndpoint) from useLoadFromEndpoint

Would it maybe make more sense to just pass the formatting-method to useLoadFromEndpoint?

File: tsconfig.json [hunk 2]
```diff
@@ -1,60 +1,21 @@
...
-    // "outFile": "./",                       /* Concatenate and emit output to single file. */
-    // "outDir": "./",                        /* Redirect output structure to the directory. */
-    // "rootDir": "./",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
-    // "removeComments": true,                /* Do not emit comments to output. */
-    "noEmit": true,                           /* Do not emit outputs. */
-    // "incremental": true,                   /* Enable incremental compilation */
-    // "importHelpers": true,                 /* Import emit helpers from 'tslib'. */
-    // "downlevelIteration": true,            /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
-    "isolatedModules": true,                  /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */
+    "target": "esnext",
```
Comment (SteffiStuffel) [inline tsconfig.json:4]: Just deleted the comments, right?

**Review by SteffiStuffel — APPROVED** [review]
Really nice PR! :tada: 

So now it is possible to either use the hook or the promise-returning-method if doing a server request? 

Thanks for adding the ts:check, had this in my mind for weeks and never did it :thinking: 

And nice to have tests now :rocket:

**Comment by SteffiStuffel** [comment]
> @SteffiStuffel it would now be really easy to replace the hoc `withCustomDisciplines` with a equivalent hook or function :) If you want I could create a draft PR for that.

I'm already working (and almost finished) with this.

**Comment by klinzo** [comment]
> So now it is possible to either use the hook or the promise-returning-method if doing a server request?
> 
Yes, it should be possible for all requests.

File: src/hooks/useLoadDocuments.ts (hunk 1 above)
Comment (klinzo) [reply src/hooks/useLoadDocuments.ts:38]: I think that is not as it is, there should be no function call of `loadDocuments` in `useLoadFromEndpoint` :thinking: So we just have dependencies from `useLoadDocuments` to `useLoadFromEndpoint` (and their other functions), so only one way.

File: tsconfig.json (hunk 2 above)
Comment (klinzo) [reply tsconfig.json:4]: I also added `skipLibCheck` and extended the `types` array with `styled-components-react-native`

Comment (klinzo) [reply tsconfig.json:4]: But yes, mostly just deleted comments.

File: src/hooks/useLoadDocuments.ts (hunk 1 above)
Comment (SteffiStuffel) [reply src/hooks/useLoadDocuments.ts:38]: yes, got this, but it gets passed to it and is then executed. My point is more about the execution order and not the dependencies.

Comment (klinzo) [reply src/hooks/useLoadDocuments.ts:38]: A I see. Yeah not sure about a better way. If we just pass the formatting method it won't be possible to use the fetch logic without hook.

**Review by Taggotty — APPROVED** [review]

**Comment by Taggotty** [comment]
Tested on Android emulator. No issues encountered: 👍🏼

**Pull request merged**